cpp/src/arrow/array/diff.cc

```diff
@@ -0,0 +1,737 @@
+      return Diff(base.begin(), base.end(), target.begin(), target.end());
+    }
+    auto base = MakeNullOrViewRange<ArrayType>(base_);
+    auto target = MakeNullOrViewRange<ArrayType>(target_);
+    return Diff(base.begin(), base.end(), target.begin(), target.end());
+  }
+
+  Status Visit(const ExtensionType&) {
+    auto base = checked_cast<const ExtensionArray&>(base_).storage();
+    auto target = checked_cast<const ExtensionArray&>(target_).storage();
+    return arrow::Diff(*base, *target, pool_, out_);
+  }
+
+  Status Visit(const DictionaryType& t) {
+    return Status::NotImplemented("diffing arrays of type ", t);
+  }
+
+  Status Diff() { return VisitTypeInline(*base_.type(), this); }
+
+  template <typename Iterator>
+  Status Diff(Iterator base_begin, Iterator base_end, Iterator target_begin,
+              Iterator target_end) {
+    QuadraticSpaceMyersDiff<Iterator> impl(base_begin, base_end, target_begin,
+                                           target_end);
+    while (!impl.Done()) {
+      impl.Next();
+    }
+    return impl.GetEdits(pool_, out_);
+  }
+
+  const Array& base_;
+  const Array& target_;
+  MemoryPool* pool_;
+  std::shared_ptr<Array>* out_;
+};
+
+Status Diff(const Array& base, const Array& target, MemoryPool* pool,
+            std::shared_ptr<Array>* out) {
+  if (!base.type()->Equals(target.type())) {
+    return Status::TypeError("only taking the diff of like-typed arrays is supported.");
+  }
+
+  return DiffImpl{base, target, pool, out}.Diff();
+}
+
+Status DiffVisitor::Visit(const Array& edits) {
+  static const auto edits_type =
+      struct_({field("insert", boolean()), field("run_length", uint64())});
+  DCHECK(edits.type()->Equals(*edits_type));
+  DCHECK_GE(edits.length(), 1);
+
+  auto insert = checked_pointer_cast<BooleanArray>(
+      checked_cast<const StructArray&>(edits).field(0));
+  auto run_lengths =
+      checked_pointer_cast<UInt64Array>(checked_cast<const StructArray&>(edits).field(1));
+
+  DCHECK(!insert->Value(0));
+
+  auto length = run_lengths->Value(0);
+  RETURN_NOT_OK(Run(length));
+
+  int64_t base_index = length, target_index = length;
+  for (int64_t i = 1; i < edits.length(); ++i) {
+    if (insert->Value(i)) {
+      RETURN_NOT_OK(Insert(target_index));
+      ++target_index;
+    } else {
+      RETURN_NOT_OK(Delete(base_index));
+      ++base_index;
+    }
+    length = run_lengths->Value(i);
+    RETURN_NOT_OK(Run(length));
+    base_index += length;
+    target_index += length;
+  }
+  return Status::OK();
+}
+
+class Formatter {
+ public:
+  void operator()(const Array& array, int64_t index, std::ostream* os) const {
+    impl_->Format(array, index, os);
+  }
+
+  static Result<Formatter> Make(const Array& arr) {
+    Formatter out;
+    RETURN_NOT_OK(VisitArrayInline(arr, &out));
+    return std::move(out);
+  }
+
+ private:
+  template <typename VISITOR>
+  friend Status VisitArrayInline(const Array& array, VISITOR* visitor);
+
+  struct Impl {
```

Comment:
Hmm... I don't understand what this is all for. Can't you use `std::function`?

Reply:
I'll replace with std::function